File: problem200.java
import java.util.LinkedList;
import java.util.Queue;

//problem 200 Number of Islands
class Solution {
    public int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0) {
            return 0;
        }
        
        int nr = grid.length;
        int nc = grid[0].length;
        int num_islands = 0;
        
        for (int r = 0; r < nr; ++r) {
            for (int c = 0; c < nc; ++c) {
                if (grid[r][c] == '1') {
                    ++num_islands;
                    grid[r][c] = '0';   //visited
                    Queue<Coordinate> neighbors = new LinkedList<>();
                    neighbors.add(new Coordinate(r,c));
                    while (!neighbors.isEmpty()) {
                        Coordinate id = neighbors.remove();
                        int row = id.x;
                        int col = id.y;
                        if (row-1 >= 0 && grid[row-1][col] == '1') {
                            neighbors.add(new Coordinate(row-1,col));
                            grid[row-1][col] = '0';
                        }
                        if (row+1 < nr && grid[row+1][col] == '1') {
                            neighbors.add(new Coordinate(row+1,col));
                            grid[row+1][col] = '0';
                        }
                        if (col-1 >= 0 && grid[row][col-1] == '1') {
                            neighbors.add(new Coordinate(row,col-1));
                            grid[row][col-1] = '0';
                        }
                        if (col+1 < nc && grid[row][col+1] == '1') {
                            neighbors.add(new Coordinate(row,col+1));
                            grid[row][col+1] = '0';
                        }
                    }
                }
            }
        }
        return num_islands;
    }
}
class Coordinate {
    int x,y;
    public Coordinate(int x, int y) {
        this.x = x;
        this.y = y;
    }
}

File: problem200.py
#problem 200 / number of islands
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0
        m, n = len(grid),len(grid[0])
        count = 0
        def bfs(i, j):
            q, grid[i][j] = [(i, j)], "0"
            for i, j in q:
                for x, y in ((i-1,j), (i+1,j), (i,j-1), (i,j+1)):
                    if 0 <= x < m and 0 <= y < n and grid[x][y] == "1":
                        grid[x][y] = "0"
                        q.append((x, y))
        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1":
                    bfs(i,j)
                    count += 1
        return count

#
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        if not grid:
            return 0
        count = 0
        neighbor = []
        m, n = len(grid),len(grid[0])
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    count += 1
                    grid[i][j] = '0'
                    neighbor.append((i,j))
                    for i,j in neighbor:
                        for x,y in ((i-1,j),(i+1,j),(i,j-1),(i,j+1)):
                            if 0 <= x < len(grid) and 0 <= y < len(grid[0]) and grid[x][y] == '1':
                                neighbor.append((x,y))
                                grid[x][y] = '0'
        return count